### ISSUE TITLE
Add file search in seafile community edition

### ISSUE THREAD

**Issue by arjones85** [haiwen/seafile-server#130]
Can we please have basic file search added to the web gui in the open source version? Don't need full text content search capabilities, but it would be really nice to at least be able to search by file name in each library on the web. This is a pretty basic feature that I think is sorely missing.

Thank you!

**Comment by hungerburg** [comment]
Search for Substring match in filenames within repository would be a welcome addition to Web API V2 too!

**Comment by mbong** [comment]
I'd even like to pay for basic file search, but the pro edition is no way out because mine is running on a Raspberry Pi and the pro edition is not available for this platform.

**Comment by ljb2of3** [comment]
Yes, please! I can definitely understand leaving full file content search to the pro edition, but searching by file name is a must! I didn't even realize this was missing until today when I suddenly needed it.

**Comment by arjones85** [comment]
Here's to hoping this next open source version will have file search in it given all the search changes in Pro that were just released!

**Comment by nan008** [comment]
We just found out there is no file search function on the community version, like seriously? This is basic functionality on EVERY cloud storage solution, so if we are trying to get people off Dropbox and told them they cannot search through their files, they laugh at me. We are not asking for in text search, just search library for file

**Comment by toplinuxsir** [comment]
Yes, Hope basic file searching will be available in the next release!

**Comment by CountCypher** [comment]
+1 for basic file search in CE. Clearly needed!

**Comment by KardinalReusen** [comment]
+1 for basic file search in CE !!!

**Comment by kostya-fr** [comment]
+1 please 🙏

**Comment by adam-ah** [comment]
@killing @freeplant If we create a PR to add file-name search to the CE edition, will you accept it?

**Comment by adam-ah** [comment]
After looking briefly into it, looks like the architecture might make it more challenging than expected. To get the list of files/dirs, the system calls deep into the object-store, vets permissions, etc. To implement search with correct permissions, one would need to either walk the whole tree (which may or may not bee too slow), or cache all the names up-front and iterate through those.

The former might work for not too large repos.

The call current call structure from front-end (python) to backend (C):

```
list_dir_with_perm -> seafile_list_dir_with_perm -> seaf_repo_manager_list_dir_with_perm -> seaf_fs_manager_get_seafdir -> seaf_obj_store_read_obj -> ...
```

**Comment by suprdave23** [comment]
> After looking briefly into it, looks like the architecture might make it more challenging than expected. To get the list of files/dirs, the system calls deep into the object-store, vets permissions, etc. To implement search with correct permissions, one would need to either walk the whole tree (which may or may not bee too slow), or cache all the names up-front and iterate through those.
> 
> The former might work for not too large repos.
> 
> The current call structure from front-end (python) to backend (C):
> 
> ```
> list_dir_with_perm -> seafile_list_dir_with_perm -> seaf_repo_manager_list_dir_with_perm -> seaf_fs_manager_get_seafdir -> seaf_obj_store_read_obj -> ...
> ```

Sounds like maybe we need a routine that writes file names to a SQL table that can then be indexed and quickly searched by the Web UI. 

Can make it an optional feature since this would expose file names in encrypted libraries.

**Comment by adam-ah** [comment]
@suprdave23 storing the filenames in SQL may or may not help - if the store access is fast enough, then searching through the filenames with patterns (ie., looking for anything containing "contract") is comparable speed in SQL and in code.
SQL, unless you are using full-text index, will perform a linear pattern matching as well, but your search terms will be tied to what SQL can do. If you ask for all file names and apply the pattern matching in code, you can apply a regex to each filename as they come in, irrespective of SQL implementation.

A further issue with dumping filenames to SQL is that if SQL names ever get out of sync (and eventually they always do), then you will also need a feature to rebuild the index. Might be more pain than gain.
